### PR TITLE
[SPARK-35063][SQL] Group exception messages in sql/catalyst

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
This PR group exception messages in `sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst`.

### Why are the changes needed?
It will largely help with standardization of error messages and its maintenance.

### Does this PR introduce _any_ user-facing change?
No. Error messages remain unchanged.

### How was this patch tested?
No new tests - pass all original tests to make sure it doesn't break any existing behavior.

### CODE REVIEW

**Comment by beliefer** [comment]
ping @allisonwang-db

**Review by allisonwang-db — COMMENTED** [review]
I checked the remaining errors under the `sql/catalyst` module and here is the list: 
(some of them are new or missed errors in already grouped sub-modules and we can skip them in this PR.)

'sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst'
| Filename                     |   Count |
|------------------------------|---------|
| CatalystTypeConverters.scala |       7 |
| JavaTypeInference.scala      |       1 |
| ScalaReflection.scala        |       6 |


'sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis' (can skip)
| Filename                          |   Count |
|-----------------------------------|---------|
| Analyzer.scala                    |       6 |
| CTESubstitution.scala             |       1 |
| CheckAnalysis.scala               |       3 |
| FunctionRegistry.scala            |       4 |
| ResolveUnion.scala                |       2 |
| TableOutputResolver.scala         |       3 |
| TypeCoercion.scala                |       2 |
| UnsupportedOperationChecker.scala |       1 |
| package.scala                     |       2 |
| v2ResolutionPlans.scala           |       2 |


'sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog' (can skip)
| Filename                    |   Count |
|-----------------------------|---------|
| ExternalCatalog.scala       |       4 |
| ExternalCatalogUtils.scala  |       2 |
| GlobalTempViewManager.scala |       1 |
| InMemoryCatalog.scala       |       9 |
| SessionCatalog.scala        |      16 |


'sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv'
| Filename               |   Count |
|------------------------|---------|
| CSVExprUtils.scala     |       4 |
| CSVHeaderChecker.scala |       1 |
| CSVInferSchema.scala   |       1 |
| CSVOptions.scala       |       4 |
| UnivocityParser.scala  |       3 |


'sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders'
| Filename                |   Count |
|-------------------------|---------|
| ExpressionEncoder.scala |       4 |
| OuterScopes.scala       |       1 |
| RowEncoder.scala        |       2 |


'sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions' (can skip)
| Filename                    |   Count |
|-----------------------------|---------|
| BoundAttribute.scala        |       1 |
| Cast.scala                  |       2 |
| ProjectionOverSchema.scala  |       2 |
| TimeWindow.scala            |       2 |
| bitwiseExpressions.scala    |       2 |
| collectionOperations.scala  |      20 |
| complexTypeCreator.scala    |       2 |
| complexTypeExtractors.scala |       2 |
| jsonExpressions.scala       |       1 |
| misc.scala                  |       4 |
| package.scala               |       1 |


'sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate'
| Filename    |   Count |
|-------------|---------|
| First.scala |       1 |


'sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen'
| Filename            |   Count |
|---------------------|---------|
| CodeGenerator.scala |       1 |


'sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects'
| Filename      |   Count |
|---------------|---------|
| objects.scala |       5 |


'sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json'
| Filename               |   Count |
|------------------------|---------|
| JacksonGenerator.scala |       3 |
| JacksonParser.scala    |       7 |
| JacksonUtils.scala     |       1 |
| JsonInferSchema.scala  |       3 |


'sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer' (can skip)
| Filename                              |   Count |
|---------------------------------------|---------|
| DecorrelateInnerQuery.scala           |       2 |
| NestedColumnAliasing.scala            |       1 |
| NormalizeFloatingNumbers.scala        |       2 |
| Optimizer.scala                       |       3 |
| PushExtraPredicateThroughJoin.scala   |       1 |
| ReplaceExceptWithFilter.scala         |       1 |
| ReplaceNullWithFalseInPredicate.scala |       1 |
| joins.scala                           |       2 |
| objects.scala                         |       3 |
| subquery.scala                        |       2 |


'sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser' (can skip)
| Filename          |   Count |
|-------------------|---------|
| AstBuilder.scala  |       4 |
| ParseDriver.scala |       2 |
| ParserUtils.scala |       1 |


'sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning'
| Filename       |   Count |
|----------------|---------|
| patterns.scala |       2 |


'sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans'
| Filename        |   Count |
|-----------------|---------|
| joinTypes.scala |       1 |


'sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical'
| Filename             |   Count |
|----------------------|---------|
| AnalysisHelper.scala |       1 |
| LogicalPlan.scala    |       1 |
| statements.scala     |       1 |


'sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation'
| Filename              |   Count |
|-----------------------|---------|
| UnionEstimation.scala |       1 |
| ValueInterval.scala   |       2 |


'sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical'
| Filename           |   Count |
|--------------------|---------|
| partitioning.scala |       1 |


'sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules'
| Filename               |   Count |
|------------------------|---------|
| RuleExecutor.scala     |       4 |
| RuleIdCollection.scala |       1 |


'sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/streaming'
| Filename                  |   Count |
|---------------------------|---------|
| InternalOutputModes.scala |       1 |


'sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees'
| Filename       |   Count |
|----------------|---------|
| TreeNode.scala |       1 |


'sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util'
| Filename                      |   Count |
|-------------------------------|---------|
| ArrayBasedMapBuilder.scala    |       1 |
| ArrayData.scala               |       2 |
| CharVarcharUtils.scala        |       1 |
| CompressionCodecs.scala       |       1 |
| DateTimeFormatterHelper.scala |       5 |
| DateTimeUtils.scala           |       1 |
| FailureSafeParser.scala       |       1 |
| IntervalUtils.scala           |       9 |
| StringUtils.scala             |       1 |
| TimestampFormatter.scala      |       1 |

**Comment by beliefer** [comment]
ping @allisonwang-db

**Comment by beliefer** [comment]
ping @allisonwang-db

**Comment by beliefer** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 1]
```diff
@@ -1451,4 +1454,146 @@ private[spark] object QueryCompilationErrors {
...
+         |variable spark.sql.crossJoin.enabled=true"""
+        .stripMargin)
+  }
+
+  def usePythonUDFInJoinConditionUnsupportedError(joinType: JoinType): Throwable = {
+    new AnalysisException("Using PythonUDF in join condition of join type" +
+      s" $joinType is not supported.")
+  }
+
+  def foundConflictAttributesInJoinConditionError(
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1573]: conflictingAttributesInJoinConditionError

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 2]
```diff
@@ -1451,4 +1454,146 @@ private[spark] object QueryCompilationErrors {
...
+      conflictingAttrs: AttributeSet, outerPlan: LogicalPlan, subplan: LogicalPlan): Throwable = {
+    new AnalysisException("Found conflicting attributes " +
+      s"${conflictingAttrs.mkString(",")} in the condition joining outer plan:\n  " +
+      s"$outerPlan\nand subplan:\n  $subplan")
+  }
+
+  def windowExprIsEmptyError(expr: Window): Throwable = {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1580]: emptyWindowExpressionError

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 3]
```diff
@@ -931,4 +937,283 @@ object QueryExecutionErrors {
...
+         |The conflict keys: ${conflictKeys.mkString(", ")}
+         |""".stripMargin)
+  }
+
+  def pairUnsupportedAtFunctionError(
+      r1: ValueInterval, r2: ValueInterval, function: String): Throwable = {
+    new UnsupportedOperationException(s"Not supported pair: $r1, $r2 at $function()")
+  }
+
+  def onceStrategyNotAllowBreakIdempotenceError[TreeType <: TreeNode[_]](
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:1175]: Maybe `onceStrategyIdempotenceIsBrokenForBatchError`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 4]
```diff
@@ -931,4 +937,283 @@ object QueryExecutionErrors {
...
+      batchName: String, plan: TreeType, reOptimized: TreeType): Throwable = {
+    new RuntimeException(
+      s"""
+         |Once strategy's idempotence is broken for batch $batchName
+         |${sideBySide(plan.treeString, reOptimized.treeString).mkString("\n")}
+       """.stripMargin)
+  }
+
+  def cannotBreakStructuralIntegrityOfInputPlanError(className: String): Throwable = {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:1184]: `structuralIntegrityOfInputPlanIsBrokenInClassError`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 5]
```diff
@@ -931,4 +937,283 @@ object QueryExecutionErrors {
...
+    new RuntimeException("The structural integrity of the input plan is broken in " +
+      s"$className.")
+  }
+
+  def cannotBreakStructuralAfterApplyRuleError(ruleName: String, batchName: String): Throwable = {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:1189]: `structuralIntegrityIsBrokenAfterApplyingRuleError`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 6]
```diff
@@ -931,4 +937,283 @@ object QueryExecutionErrors {
...
+    new RuntimeException(s"After applying rule $ruleName in batch $batchName, " +
+      "the structural integrity of the plan is broken.")
+  }
+
+  def notFoundRuleIdForRuleError(ruleName: String): Throwable = {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:1194]: ruleIdNotFoundForRuleError

**Review by allisonwang-db — APPROVED** [review]
Looks good! Added a few comments.

**Comment by beliefer** [comment]
cc @cloud-fan

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by beliefer** [comment]
@allisonwang-db Thanks you for your review
@cloud-fan Thank you too.

**Pull request closed without merging**